Follow up #29365, fix fail to remove container after restart

coolljt0725 opened the pull request:
Call daemon.Mount will increase the refcount of mounted layer,
for those previous running containers, `Mount` call will make
the refcount to 2. see
https://github.com/docker/docker/blob/v1.13.0-rc4/daemon/graphdriver/counter.go#L38
```
	if !m.check {
		m.check = true
		if c.checker.IsMounted(path) {
			m.count++
		}
	}
	m.count++

```
graphdrive could restore on reboot after #22541, call
daemon.Mount to restore the graphdriver is not necessary.

And call `daemon.Mount` on restoring will mount all the containers
mounted layer even if it was stop.

````
[lei@centos docker]$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
[lei@centos docker]$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                     PORTS               NAMES
54fb9a9e08fe        662917bef212        "sh"                4 days ago          Exited (0) 4 days ago                          keen_archimedes
c24d30fa87bc        662917bef212        "bash"              4 days ago          Created                                        peaceful_mahavira
b78fb816e787        busybox             "sh"                5 days ago          Exited (0) 5 days ago                          loving_lamarr
7d705bb81a63        busybox             "sh"                5 days ago          Exited (0) 5 days ago                          tender_wiles
0611718b63e7        busybox             "sh"                3 weeks ago         Exited (255) 5 days ago                        lucid_heisenberg
30b80e1ff02c        busybox             "sh"                3 weeks ago         Exited (137) 3 weeks ago                       keen_newton
a5f64b4c78ef        busybox             "sh"                3 weeks ago         Exited (137) 3 weeks ago                       objective_wing
090bd488a48b        busybox             "sh"                3 weeks ago         Exited (130) 3 weeks ago                       gallant_lamport
431573f53e87        busybox             "sh"                3 weeks ago         Exited (130) 3 weeks ago                       infallible_sammet
[lei@centos docker]$ mount | grep devicemapper
/dev/mapper/centos-root on /var/lib/docker/devicemapper type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/mapper/docker-253:0-135201377-e9b3ff54b53225c8ae8ed1b5dd300f0c790d95d80d7450fed1cb912fa9e6aae4 on /var/lib/docker/devicemapper/mnt/e9b3ff54b53225c8ae8ed1b5dd300f0c790d95d80d7450fed1cb912fa9e6aae4 type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-81a07eb76ac0e0ca8868e8ad8abf39e03bee1ff182b00086c49be2f33aa51fed on /var/lib/docker/devicemapper/mnt/81a07eb76ac0e0ca8868e8ad8abf39e03bee1ff182b00086c49be2f33aa51fed type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-c38d85dfb9de2a4437f96240f717b2f855fe12c3d35ae3aa574633d7a3eb3b18 on /var/lib/docker/devicemapper/mnt/c38d85dfb9de2a4437f96240f717b2f855fe12c3d35ae3aa574633d7a3eb3b18 type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-741c2a790ea7a48dadbabbc3ba74c0ed58d6d8d141ea091c43f39737ce55a309 on /var/lib/docker/devicemapper/mnt/741c2a790ea7a48dadbabbc3ba74c0ed58d6d8d141ea091c43f39737ce55a309 type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-7b3f2c52b6e11f9f1d5e9833b585e96d38d19af19f80437bb94ba010bd18bf28 on /var/lib/docker/devicemapper/mnt/7b3f2c52b6e11f9f1d5e9833b585e96d38d19af19f80437bb94ba010bd18bf28 type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-7acdbd273d5ccf104018c222982776361c4d34b5155cad965fb4b8ac2d1a9e3a on /var/lib/docker/devicemapper/mnt/7acdbd273d5ccf104018c222982776361c4d34b5155cad965fb4b8ac2d1a9e3a type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-cee4d9bb8de978ae6c04a2f2406e9fec1405a0d18ee22abd873c733bddfac401 on /var/lib/docker/devicemapper/mnt/cee4d9bb8de978ae6c04a2f2406e9fec1405a0d18ee22abd873c733bddfac401 type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-99cfd02ecb0e73b50a13846b1d9432a26ea93424ecf037ca028cc8a8a12a38ce on /var/lib/docker/devicemapper/mnt/99cfd02ecb0e73b50a13846b1d9432a26ea93424ecf037ca028cc8a8a12a38ce type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)
/dev/mapper/docker-253:0-135201377-7caf7291e38fbf7b0880556ad61408c5ed916a3833899ead9091cfd835580147 on /var/lib/docker/devicemapper/mnt/7caf7291e38fbf7b0880556ad61408c5ed916a3833899ead9091cfd835580147 type xfs (rw,relatime,seclabel,nouuid,attr2,inode64,logbsize=64k,sunit=128,swidth=128,noquota)

````

and if you use `devicemapper`, you will failed to remove container after restarting
```
[lei@centos docker]$ docker rm 7d705bb81a63
Error response from daemon: Driver devicemapper failed to remove root filesystem 7d705bb81a63282c7e2bcfe76140617670c19a909049c60678bd438be5c64d08: Device is Busy
```

This fix call Mount and then Unmount to get `BaseFs`

ping @yongtang  @vdemeester  @mlaventure @cpuguy83  @tonistiigi 

Signed-off-by: Lei Jitang <leijitang@huawei.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

yongtang commented:
@coolljt0725 Sorry for missing the unmount part. And thanks for taking care of this issue 👍

coolljt0725 commented:
@yongtang  It's ok, sorry for our bad review :)

thaJeztah commented:
ping @mlaventure PTAL

tonistiigi commented:
LGTM

@coolljt0725 Could we add a test for this? Maybe a separate daemon that always runs overlay graphdriver.

coolljt0725 commented:
@tonistiigi  test added, but this need run on overlay daemon and set `DOCKER_GRAPHDRIVER=overlay`  to run the integration test

tonistiigi commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -2848,3 +2849,36 @@ func (s *DockerDaemonSuite) TestExecWithUserAfterLiveRestore(c *check.C) {
+
+func (s *DockerDaemonSuite) TestRemoveContainerAfterLiveRestore(c *check.C) {
+	testRequires(c, DaemonIsLinux, IsOverlay)
+	s.d.StartWithBusybox(c, "--live-restore", "-s", "overlay")
```
I think you need to use `--storage-driver`. https://github.com/docker/docker/blob/bf6eb855391e166a65db284567686fafbe4a9a0a/integration-cli/daemon/daemon.go#L243-L245

coolljt0725 replied:
updated

coolljt0725 commented:
test locally, with the fix
```
INFO: Testing against a local daemon
PASS: docker_cli_daemon_test.go:2191: DockerDaemonSuite.TestRemoveContainerAfterLiveRestore     2.746s
OK: 1 passed
PASS
coverage: 20.9% of statements
ok      github.com/docker/docker/integration-cli        17.919s
---> Making bundle: .integration-daemon-stop (in bundles/1.11.2/test-integration-cli)
++++ cat bundles/1.11.2/test-integration-cli/docker.pid
+++ kill 9533

```
without the fix
```
+ go test -check.f TestRemoveContainerAfterLiveRestore -check.v -check.timeout=5m -timeout=360m github.com/docker/docker/integration-cli
L: docker_cli_daemon_test.go:2191: DockerDaemonSuite.TestRemoveContainerAfterLiveRestore

[d11061248] waiting for daemon to start
[d11061248] daemon started
[d11061248] exiting daemon
[d11061248] waiting for daemon to start
[d11061248] daemon started
docker_cli_daemon_test.go:2215:
    c.Fatalf("mountinfo should not include the mountpoint of stop container")
... Error: mountinfo should not include the mountpoint of stop container

timeout
Attempt #2: daemon is still running with pid 18412
[d11061248] exiting daemon
OOPS: 0 passed, 1 FAILED
--- FAIL: Test (18.34s)
FAIL
coverage: 20.9% of statements
exit status 1
FAIL    github.com/docker/docker/integration-cli        18.417s
---> Making bundle: .integration-daemon-stop (in bundles/1.11.2/test-integration-cli)
+++++ cat bundles/1.11.2/test-integration-cli/docker.pid
++++ kill 15717

```

tonistiigi commented:
LGTM, Thanks @coolljt0725 !

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -213,6 +204,21 @@ func (daemon *Daemon) restore() error {
 				}
+
+				// we call Mount and then Unmount to get BaseFs of the container
+				if err := daemon.Mount(c); err != nil {
```
This will fail if it's already mounted though, right?
Can we check if it's already mounted before logging an error?
We have too many mount related warning messages in the daemon logs as is.

tonistiigi replied:
@cpuguy83 It shouldn't. We call it all the time for cp actions for example. It just increases the refcount.

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -213,6 +204,21 @@ func (daemon *Daemon) restore() error {
+					// The error is only logged here.
+					logrus.Warnf("Failed to mount container on getting BaseFs path %v: %v", c.ID, err)
+				}
+				if err := daemon.Unmount(c); err != nil {
```
Likewise maybe we can check if it's (not) mounted before logging a warning about it.

coolljt0725 replied:
@cpuguy83  How about this 
```
if err := daemon.Mount(c); err != nil {
    log.Warnf()
} else {
    if err := daemon.Unmount(c); err != nil {
    log.Warnf()
}
}
```
so the unmount will only call when mount call success.

cpuguy83 replied:
I guess this makes sense, if mount failed we don't want to unmount and mess with ref counts.

cpuguy83 commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -2848,3 +2849,36 @@ func (s *DockerDaemonSuite) TestExecWithUserAfterLiveRestore(c *check.C) {
 }
+
+func (s *DockerDaemonSuite) TestRemoveContainerAfterLiveRestore(c *check.C) {
+	testRequires(c, DaemonIsLinux, IsOverlay)
```
Isn't `IsOverlay` redundant since you are just setting overlay below?
Maybe `overlayFSSuported` instead?

cpuguy83 replied:
Also needs `SameHostDaedmon`

coolljt0725 replied:
@cpuguy83  I use `IsOverlay` to make sure the daemon is support overlay, using `overlayFSSuported` is also a good idea,

coolljt0725 commented:
update with @cpuguy83  's suggestions

coolljt0725 committed:
Follow up #29365, fix fail to remove container after restart

Call daemon.Mount will increase the refcount of mounted path,
for those previous running containers, `Mount` call will make
the refcount to 2. see
https://github.com/docker/docker/blob/v1.13.0-rc4/daemon/graphdriver/counter.go#L38
```
	if !m.check {
		m.check = true
		if c.checker.IsMounted(path) {
			m.count++
		}
	}
	m.count++

```
graphdrive could restore on reboot after #22541, call
daemon.Mount to resore the graphdriver is not necessary.

And call `daemon.Mount` on restorting will mount all the containers
mounted layer even if it was stop.

This fix call Mount and then Unmount to get `BaseFs`

Signed-off-by: Lei Jitang <leijitang@huawei.com>

cpuguy83 reviewed (approved):
LGTM

thaJeztah commented:
all green now
